overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,231 @@
+            sourceArn: bucket.id.apply(bucketName => `arn:aws:s3:::${bucketName}`),
+        }, { parent: this });
+
+        this.permission = permission;
+
+        // We must create only a single BucketNotification per Bucket per AWS API limitations.  See
+        // https://github.com/terraform-providers/terraform-provider-aws/issues/1715.  So we push
+        // the subscription information here, and then actually create the BucketNotification if
+        // needed on process `beforeExit`.
+        let subscriptions = bucketSubscriptionInfos.get(bucket);
+        if (!subscriptions) {
+            subscriptions = [];
+            bucketSubscriptionInfos.set(bucket, subscriptions);
+        }
+
+        subscriptions.push({
+            name: name,
+            events: args.events,
+            filterPrefix: args.filterPrefix,
+            filterSuffix: args.filterSuffix,
+            lambdaFunctionArn: func.arn,
+            permission: permission,
+        });
+    }
+}
+
+process.on("beforeExit", () => {
+    const copy = bucketSubscriptionInfos;
+
+    // Since we are generating more work on the event loop, we will casue `beforeExit` to be invoked again.
+    // Make sure to clear out eh pending subscrpitions array so that we don't try to apply them again.
+    bucketSubscriptionInfos = new Map();
+
+    for (const [bucket, subscriptions] of copy) {
+        const permissions = subscriptions.map(s => s.permission);
+        const _ = new BucketNotification(subscriptions[0].name, {
+            bucket: bucket.id,
+            lambdaFunctions: subscriptions.map(subscription => ({
+                events: subscription.events,
+                filterPrefix: subscription.filterPrefix,
+                filterSuffix: subscription.filterSuffix,
+                lambdaFunctionArn: subscription.lambdaFunctionArn,
+            })),
+        }, { parent: bucket, dependsOn: permissions });
+    }
+});
+
+// Mixin event handling functionality onto Bucket.
+declare module "./bucket" {
+    interface Bucket {
+        onObjectCreated(
+            name: string, handler: BucketEventHandler,
+            args?: ObjectCreatedSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
+
+        onObjectRemoved(
+            name: string, handler: BucketEventHandler,
+            args?: ObjectRemovedSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
+
+        /**
+         * Creates a new subscription to the given bucket using the lambda provided, along with
+         * optional options to control the behavior of the subscription.  This function should be
+         * used when full control over the subscription is wanted, and other helpers (like
+         * onObjectCreated/onObjectRemoved) are not sufficient.
+         */
+        onEvent(
+            name: string, handler: BucketEventHandler,
+            args: BucketEventSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
```

Comment:
this is also a part of our opinionated view on events.  In general, most resources will have a generic onEvent entrypoint that gives them full control over whatever AWS exposes.  For example, you may want to specifically list to ObjectCreated:Post vs ObjectCreated:Put.  Or you might want a single event that hears about both creation and deletes.  These onEvent helpers try to expose the full subscription functionality.  

In general, we will try to provide other suitable helpers (like onObjectCreated and onObjectDeleted) when AWS has high level event bucketing themselve.  These helpers should almost always call into the more generic onEvent entrypoint, just with the right values set for the user.

Reply:
Fundamentally though, that means that all these helpers may look very different across cloud providers.  The Azure versions here will have a very different naming/shape.  Unification (if any) happens at the @pulumi/cloud layer.